### PR TITLE
Vp8_enc: Implement VP8 SVC-T encoding

### PR DESCRIPTION
**Pull request by wudping**
Add an interface IVaapiFlagParameter that contains all the common functions for non-SVC-T and SVC-T.
Class VaapiFlagParameterNormal to implement picture parameters setting in non-SVC-T mode.
Class VaapiFlagParameterSVCT in SVC-T mode.

### CODE REVIEW

**Comment by wudping** [comment]
please help to review. Thanks.

**Comment by uartie** [comment]
Please fix build error

**Comment by wudping** [comment]
Hi @uartie , 
The reason for 9495510ef698 compilation failure, is that the commit applies a variable "temporal_id” that still hasn't been merged into current Libva Master branch. My local Libva has the variable, so it can pass through the compilation. 

Error info:
dec::VaapiEncoderVP8::fill(VAEncPictureParameterBufferVP8*, const PicturePtr&, const SurfacePtr&) const':
13:03:48,452 INFO  - vaapiencoder_vp8.cpp:497:30: error: 'struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>**' has no member named 'temporal_id'**
13:03:48,452 INFO  - **picParam->ref_flags.bits.temporal_id** = picture->m_temporalLayerID;

File: encoder/vaapiencoder_vp8.cpp [hunk 1]
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
     }
 };
 
-VaapiEncoderVP8::VaapiEncoderVP8():
-	m_frameCount(0),
-	m_qIndex(VP8_DEFAULT_QP)
+class IVaapiFlagParameter {
```
Comment (lizhong1008) [inline encoder/vaapiencoder_vp8.cpp]: What does flag parameter mean? Do you mean FillParameter?

File: encoder/vaapiencoder_vp8.cpp [hunk 2]
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
...
+
+private:
+    VideoFrameRate m_framerates[VP8_MAX_TEMPORAL_LAYER_NUM];
+    uint32_t m_framerateRatio[VP8_MAX_TEMPORAL_LAYER_NUM];
+    uint32_t m_layerBitRate[VP8_MAX_TEMPORAL_LAYER_NUM];
+    uint32_t m_periodicity;
+    Uint32Vector m_tempLayerIDs;
+    uint8_t m_layerNum;
+};
+
```
Comment (lizhong1008) [inline encoder/vaapiencoder_vp8.cpp]: Why put the class declaration in the cpp file instead of hpp file?

File: encoder/vaapiencoder_vp8.cpp [hunk 3]
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
...
+void VaapiFlagParameterSVCT::printRatio()
+{
+    printf("ratio: \n");
+    for (uint8_t i = 0; i < m_layerNum; i++) {
+        if (i != 0)
+            printf(" : ");
+        printf("%d", m_framerateRatio[i]);
+    }
+    printf("\n");
```
Comment (lizhong1008) [inline encoder/vaapiencoder_vp8.cpp]: Please change it to DEBUG instead of print

Comment (lizhong1008) [reply encoder/vaapiencoder_vp8.cpp]: All print() functions should be replaced to DEBUG()

File: interface/VideoEncoderDefs.h [hunk 4]
```diff
@@ -152,6 +153,11 @@ typedef struct VideoFrameRate {
     uint32_t frameRateDenom;
 }VideoFrameRate;
 
+typedef struct SVCTVideoFrameRate {
+    VideoFrameRate fraction[32];
+    uint8_t num;
```
Comment (lizhong1008) [inline interface/VideoEncoderDefs.h]: Here is the interface to user, would better add some comment to describe what does fraction and num mean.

**Comment by uartie** [comment]
@wudping please provide a link to the libva PR that this depends on. thanks.

File: encoder/vaapiencoder_vp8.cpp (hunk 1 above)
Comment (wudping) [reply encoder/vaapiencoder_vp8.cpp]: It means the flags in the parameter. I can't figure out a proper name for that.

File: encoder/vaapiencoder_vp8.cpp (hunk 2 above)
Comment (wudping) [reply encoder/vaapiencoder_vp8.cpp]: Because it just for vp8 encode, not for vp9 or any other codecs.

**Comment by wudping** [comment]
hi, @uartie , 
The Libva code is on the Teamforge:
**url:** ssh://dwux@git-amr-3.devtools.intel.com:29418/otc_media-libva.git
**branch:** vp8-svct

File: encoder/vaapiencoder_vp8.cpp (hunk 2 above)
Comment (lizhong1008) [reply encoder/vaapiencoder_vp8.cpp]: you can move it to vaapiencoder_vp8.h

**Comment by lizhong1008** [comment]
@wudping please make sure "picParam->ref_flags.bits.temporal_id" is necessary for vp8 svc-t decoder and VAAPI driver or not. If it is not necessary, you can remove the line "picParam->ref_flags.bits.temporal_id = picture->m_temporalLayerID;". Then it can be buildable using upstream VAAPI.

**Comment by lizhong1008** [comment]
@wudping ,if it is necessary, you can send a pull request to 01.org/libva.

File: encoder/vaapiencoder_vp8.cpp (hunk 2 above)
Comment (xuguangxin) [reply encoder/vaapiencoder_vp8.cpp]: Hi @lizhong1008 ,
The class only used by this .cpp, it's no need add this to .h.

File: interface/VideoEncoderDefs.h (hunk 4 above)
Comment (xuguangxin) [reply interface/VideoEncoderDefs.h]: Hi @wudping , why we need this?, why h264 not needed it?

**Comment by xuguangxin** [comment]
@wudping will take some vacation,  I will help address the issue and refact the code

**Comment by xuguangxin** [comment]
I am still working on this, I hope we can refact h264 code make the svct share same code with vp8 svct. Hope I can send h264 refact code in following days

**Comment by wudping** [comment]
compile error as following is due to the libva master dose not contain the variable:

vaapiencoder_vp8.cpp:311:34: error: ‘struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>’ has no member named ‘first_ref’
         **picParam->ref_flags.bits.first_ref = 0x01;
         picParam->ref_flags.bits.second_ref = 0x02;**
                                  ^
...
vaapiencoder_vp8.cpp:335:30: error: ‘struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>’ has no member named ‘temporal_id’
     **picParam->ref_flags.bits.temporal_id = picture->m_temporalID;**

**Review by xuguangxin — APPROVED** [review]

**Comment by xuguangxin** [comment]
could you protect this with some libva version check?

```
compile error as following is due to the libva master dose not contain the variable:

vaapiencoder_vp8.cpp:311:34: error: ‘struct _VAEncPictureParameterBufferVP8::::’ has no member named ‘first_ref’
picParam->ref_flags.bits.first_ref = 0x01;
picParam->ref_flags.bits.second_ref = 0x02;
^
...
vaapiencoder_vp8.cpp:335:30: error: ‘struct _VAEncPictureParameterBufferVP8::::’ has no member named ‘temporal_id’
picParam->ref_flags.bits.temporal_id = picture->m_temporalID;

```

**Comment by wudping** [comment]
updated, thanks.

**Pull request merged**